<!DOCTYPE html>
<html lang="en-US">
  <head>
    <meta charset="UTF-8">
    <meta http-equiv="X-UA-Compatible" content="IE=edge">
    <meta name="viewport" content="width=device-width, initial-scale=1">

<!-- Begin Jekyll SEO tag v2.6.1 -->
<title>demo-data | cellxgene</title>
<meta name="generator" content="Jekyll v3.8.7" />
<meta property="og:title" content="demo-data" />
<meta property="og:locale" content="en_US" />
<meta name="description" content="Demo datasets" />
<meta property="og:description" content="Demo datasets" />
<link rel="canonical" href="https://chanzuckerberg.github.io/cellxgene/posts/launch.html" />
<meta property="og:url" content="https://chanzuckerberg.github.io/cellxgene/posts/launch.html" />
<meta property="og:site_name" content="cellxgene" />
<script type="application/ld+json">
{"@type":"WebPage","publisher":{"@type":"Organization","logo":{"@type":"ImageObject","url":"https://chanzuckerberg.github.io/cellxgene/cellxgene-logo.png"}},"headline":"demo-data","description":"Demo datasets","url":"https://chanzuckerberg.github.io/cellxgene/posts/launch.html","@context":"https://schema.org"}</script>
<!-- End Jekyll SEO tag -->

    <link rel="stylesheet" href="/cellxgene/assets/css/style.css?v=5d884a7b76632b096dce67de9fc17280f6712926">
    <!--[if lt IE 9]>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/html5shiv/3.7.3/html5shiv.min.js"></script>
    <![endif]-->
  </head>
  <body>
    <div class="wrapper">
      <header>
        <img src="/cellxgene/cellxgene-logo.png" alt="cellxgene" />

        <p>An interactive explorer for single-cell transcriptomics data</p>
        <p>

        
        <a href="/cellxgene/" class="btn">Quick start</a><br>
        

        
        
        <a href="/cellxgene/posts/install" class="btn">Installation</a><br>
        
        
        
        <a href="/cellxgene/posts/gallery" class="btn">Gallery</a><br>
        
        
        
        <a href="/cellxgene/posts/demo-data" class="btn">Demo datasets</a><br>
        
        
        
        <a href="/cellxgene/posts/prepare" class="btn">Preparing your data</a><br>
        
        
        
        <a href="/cellxgene/posts/launch" class="btn"><b>Launching cellxgene</b></a><br>
        
        
        
        <a href="/cellxgene/posts/hosted" class="btn">Hosting cellxgene</a><br>
        
        
        
        <a href="/cellxgene/posts/annotations" class="btn">Annotating data</a><br>
        
        
        
        <a href="/cellxgene/posts/methods" class="btn">Methods</a><br>
        
        
        
        <a href="/cellxgene/posts/troubleshooting" class="btn">Troubleshooting</a><br>
        
        
        
        <a href="/cellxgene/posts/roadmap" class="btn">Roadmap</a><br>
        
        
        
        <a href="/cellxgene/posts/contribute" class="btn">Contributing (ideas or code)</a><br>
        
        
        
        <a href="/cellxgene/posts/extensions" class="btn">cellxgene Extensions</a><br>
        
        
        
        <a href="/cellxgene/posts/contact" class="btn">Contact & finding help</a><br>
        
        
        
        <a href="/cellxgene/posts/cellxgene_cziscience_com" class="btn">cellxgene.cziscience.com</a><br>
        
        

        <a href="https://github.com/chanzuckerberg/cellxgene" class="btn" target="_blank">Code</a>
    </p>
    </header>

      <section>
      <h1 id="launching-cellxgene-with-your-dataset">Launching cellxgene with your dataset</h1>

<h2 id="quickstart">Quickstart</h2>

<p>Once you’ve <a href="prepare">prepared your data</a> for cellxgene, you can launch the app using</p>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>cellxgene launch mydataset.h5ad --open
</code></pre></div></div>

<p>You should see your web browser open with the following</p>

<p><img width="600" src="../images/cellxgene-opening-screenshot.png" pad="50px" /></p>

<p><strong>Note</strong>: automatic opening of the browser with the <code class="language-plaintext highlighter-rouge">--open</code> flag only works on some platforms (eg, OSX). On other platforms you’ll need to directly point to the provided link in your browser.</p>

<h2 id="launching-from-a-url">Launching from a URL</h2>

<p>You can also launch from a URL directly like this:</p>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>cellxgene launch https://github.com/chanzuckerberg/cellxgene/blob/main/example-dataset/pbmc3k.h5ad
</code></pre></div></div>

<p>Support for S3 and GCS is not enabled by default. If you wish to directly access S3 or GFS, install one or both of the following packages:</p>

<ul>
  <li><a href="https://s3fs.readthedocs.io/en/latest/">s3fs</a> for S3 support</li>
  <li><a href="https://gcsfs.readthedocs.io/en/latest/">gcsfs</a> for GCS support</li>
</ul>

<p>For example:</p>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>pip install s3fs
cellxgene launch s3://mybucket.s3-us-west-2.amazonaws.com/mydata.h5ad
</code></pre></div></div>

<h2 id="options-for-cellxgene-launch">Options for cellxgene <code class="language-plaintext highlighter-rouge">launch</code></h2>

<p><strong>For the most up-to-date and comprehensive list of options, run <code class="language-plaintext highlighter-rouge">cellxgene launch --help</code></strong></p>

<p><code class="language-plaintext highlighter-rouge">--open</code> automatically opens the web browser after launching (caveat: only works on some operating systems).</p>

<p><code class="language-plaintext highlighter-rouge">--disable-annotations</code>, <code class="language-plaintext highlighter-rouge">--annotations-file</code> &amp; <code class="language-plaintext highlighter-rouge">--annotations-dir</code> all have to do with creating new categorical annotations in the application. We have a <a href="annotations">whole separate page</a> about their usage! :)</p>

<p><code class="language-plaintext highlighter-rouge">--diffexp-lfc-cutoff</code> as explained <a href="methods">in the methods</a>, genes are only returned in differential expression if the effect size is above the specified threshold for log fold change. Defaults to 0.01.</p>

<p><code class="language-plaintext highlighter-rouge">--disable-diffexp</code> will disable and hide the <code class="language-plaintext highlighter-rouge">Compute Differential Expression</code> feature.
For large datasets, or datasets loaded with the <code class="language-plaintext highlighter-rouge">--backed</code> option, computing differential expression may
be extremely slow or use excessive resources on the host computer (e.g., memory thrashing).
Disabling the feature will ensure that this computation is not initiated accidentally.</p>

<p><code class="language-plaintext highlighter-rouge">--backed</code> option instructs <code class="language-plaintext highlighter-rouge">cellxgene launch</code> to read the H5AD file in “backed” mode (for more information, see the
<a href="https://anndata.readthedocs.io/en/latest/anndata.read_h5ad.html#anndata.read_h5ad">anndata.read_h5ad() documentation</a>).</p>

<p>By default, cellxgene will read the entire H5AD will be into memory at startup, improving application speed and performance.
Very large datasets may not fit in memory. The “–backed” mode will read the file incrementally, reducing memory
use, and for large files, improving startup speed. <em>However</em>, this option will also significantly slow
down access to gene expression histograms, and may render differential expression calculations too slow
to use (see <code class="language-plaintext highlighter-rouge">--disable-diffexp</code> for an option to disable this feature).</p>

<p><code class="language-plaintext highlighter-rouge">--embedding</code> restricts which embeddings will be available in the viewer. By default, all embeddings specified in <code class="language-plaintext highlighter-rouge">anndata.obsm['X_name']</code> will be loaded; if you have many embeddings, you may wish to restrict this list for a speedier launch.</p>

<p><code class="language-plaintext highlighter-rouge">--title</code> adds a title to the viewer. Defaults to file name.</p>

<p><code class="language-plaintext highlighter-rouge">--about</code> adds a link where users can go to find more infomation about the dataset. Requires <code class="language-plaintext highlighter-rouge">https</code>.</p>

<p><code class="language-plaintext highlighter-rouge">--obs-names</code> allows you to specify which column in <code class="language-plaintext highlighter-rouge">anndata.obs</code> to use as <code class="language-plaintext highlighter-rouge">anndata.obs.index</code>.</p>

<p><code class="language-plaintext highlighter-rouge">--var-names</code> allows you to specify which column in <code class="language-plaintext highlighter-rouge">anndata.var</code> to use as <code class="language-plaintext highlighter-rouge">anndata.var.index</code>.</p>

<p><code class="language-plaintext highlighter-rouge">--max-category-items</code> omits categorical metadata fields that contain more than N <em>distinct</em> values. Defaults to 1000.</p>

      </section>
      <footer>
        
        <p>This project is maintained by <a href="https://github.com/chanzuckerberg">chanzuckerberg</a></p>
        
      </footer>
    </div>
    <script src="/cellxgene/assets/js/scale.fix.js"></script>
    
  </body>
</html>
