<!DOCTYPE html>
<html lang="en-US">
  <head>
    <meta charset="UTF-8">
    <meta http-equiv="X-UA-Compatible" content="IE=edge">
    <meta name="viewport" content="width=device-width, initial-scale=1">

<!-- Begin Jekyll SEO tag v2.5.0 -->
<title>Methods | cellxgene</title>
<meta name="generator" content="Jekyll v3.8.5" />
<meta property="og:title" content="Methods" />
<meta property="og:locale" content="en_US" />
<link rel="canonical" href="http://localhost:4000/posts/methods.html" />
<meta property="og:url" content="http://localhost:4000/posts/methods.html" />
<meta property="og:site_name" content="cellxgene" />
<script type="application/ld+json">
{"publisher":{"@type":"Organization","logo":{"@type":"ImageObject","url":"http://localhost:4000/cellxgene-logo.png"}},"@type":"WebPage","url":"http://localhost:4000/posts/methods.html","headline":"Methods","@context":"http://schema.org"}</script>
<!-- End Jekyll SEO tag -->

    <link rel="stylesheet" href="/assets/css/style.css?v=be9a415a4a06db270354d81837f6e78c71555064">
    <!--[if lt IE 9]>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/html5shiv/3.7.3/html5shiv.min.js"></script>
    <![endif]-->
  </head>
  <body>
    <div class="wrapper">
      <header>
        <img src="/cellxgene-logo.png" alt="cellxgene" />

        <p></p>
        <p>

        
        <a href="/" class="btn">Quick start</a><br>
        

        
        
        <a href="/posts/install" class="btn">Installation</a><br>
        
        
        
        <a href="/posts/gallery" class="btn">Gallery</a><br>
        
        
        
        <a href="/posts/demo-data" class="btn">Demo datasets</a><br>
        
        
        
        <a href="/posts/prepare" class="btn">Preparing your data</a><br>
        
        
        
        <a href="/posts/launch" class="btn">Launching cellxgene</a><br>
        
        
        
        <a href="/posts/hosted" class="btn">Hosting cellxgene</a><br>
        
        
        
        <a href="/posts/annotations" class="btn">Annotating data</a><br>
        
        
        
        <a href="/posts/methods" class="btn"><b>Methods</b></a><br>
        
        
        
        <a href="/posts/troubleshooting" class="btn">Troubleshooting</a><br>
        
        
        
        <a href="/posts/roadmap" class="btn">Roadmap</a><br>
        
        
        
        <a href="/posts/contribute" class="btn">Contributing (ideas or code)</a><br>
        
        
        
        <a href="/posts/contact" class="btn">Contact & finding help</a><br>
        
        

        <a href="https://github.com/chanzuckerberg/cellxgene" class="btn" target="_blank">Code</a>
    </p>
    </header>

      <section>
      <h1 id="methods">Methods</h1>

<h2 id="data-structure-anndata-fields-used-for-visualization">Data structure: <em>anndata</em> fields used for visualization</h2>

<h3 id="expression-data">Expression data</h3>

<p>Gene expression values are pulled from <code class="highlighter-rouge">anndata.X</code>. These feed into the histograms, scatterplot, colorscale, and differential expression calculations. We’re <a href="https://github.com/chanzuckerberg/cellxgene/issues/689">working on ways</a> to incorporate <code class="highlighter-rouge">anndata.raw</code> and other <code class="highlighter-rouge">anndata.layers</code>!</p>

<h3 id="metadata">Metadata</h3>

<p>Categorical (e.g., cluster labels) and continuous (e.g., pseudotime) metadata are pulled from <code class="highlighter-rouge">anndata.obs</code>. Any column added here will be available for visualization in cellxgene. You can also <a href="annotations">create new categorical annotations</a> within the application.</p>

<h3 id="embeddings">Embeddings</h3>

<p>cellxgene looks for embeddings (e.g., tSNE, UMAP, PCA, spatial coordinates) in <code class="highlighter-rouge">anndata.obsm</code>. These fields must follow the scanpy convention of starting with <code class="highlighter-rouge">X_</code>, e.g., <code class="highlighter-rouge">anndata.obsm['X_umap']</code>. If an embedding has more than two components, the first two will be used for visualization.</p>

<h2 id="differential-expression">Differential expression</h2>

<p>We’re actively working on how to improve differential expression within the app.</p>

<p>Currently, we use a <a href="https://en.wikipedia.org/wiki/Welch%27s_t-test">Welch’s <em>t</em>-test</a> implementation, including the same variance overestimation correction as used in <code class="highlighter-rouge">scanpy</code>. We sort the <code class="highlighter-rouge">tscore</code> to identify the top 15 genes, and then filter to remove any that fall below a cutoff log fold change value, which can help remove spurious test results. The default threshold is <code class="highlighter-rouge">0.01</code> and can be changed using the option <code class="highlighter-rouge">--diffexp-lfc-cutoff</code>.</p>

      </section>
      <footer>
        
        <p>This project is maintained by <a href="https://github.com/chanzuckerberg">chanzuckerberg</a></p>
        
      </footer>
    </div>
    <script src="/assets/js/scale.fix.js"></script>
    
  </body>
</html>
