<!DOCTYPE html>
<html lang="en-US">
  <head>
    <meta charset="UTF-8">
    <meta http-equiv="X-UA-Compatible" content="IE=edge">
    <meta name="viewport" content="width=device-width, initial-scale=1">

<!-- Begin Jekyll SEO tag v2.5.0 -->
<title>Troubleshooting | cellxgene</title>
<meta name="generator" content="Jekyll v3.8.5" />
<meta property="og:title" content="Troubleshooting" />
<meta property="og:locale" content="en_US" />
<meta name="description" content="Troubleshooting" />
<meta property="og:description" content="Troubleshooting" />
<link rel="canonical" href="https://chanzuckerberg.github.io/cellxgene/posts/troubleshooting.html" />
<meta property="og:url" content="https://chanzuckerberg.github.io/cellxgene/posts/troubleshooting.html" />
<meta property="og:site_name" content="cellxgene" />
<script type="application/ld+json">
{"@type":"WebPage","headline":"Troubleshooting","publisher":{"@type":"Organization","logo":{"@type":"ImageObject","url":"https://chanzuckerberg.github.io/cellxgene/cellxgene-logo.png"}},"description":"Troubleshooting","url":"https://chanzuckerberg.github.io/cellxgene/posts/troubleshooting.html","@context":"http://schema.org"}</script>
<!-- End Jekyll SEO tag -->

    <link rel="stylesheet" href="/cellxgene/assets/css/style.css?v=b5ba39592c9370d8c06398ca9ca12e7dc89afa50">
    <!--[if lt IE 9]>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/html5shiv/3.7.3/html5shiv.min.js"></script>
    <![endif]-->
  </head>
  <body>
    <div class="wrapper">
      <header>
        <img src="/cellxgene/cellxgene-logo.png" alt="cellxgene" />

        <p>An interactive explorer for single-cell transcriptomics data</p>
        <p>

        
        <a href="/cellxgene/" class="btn">Quick start</a><br>
        

        
        
        <a href="/cellxgene/posts/install" class="btn">Installation</a><br>
        
        
        
        <a href="/cellxgene/posts/gallery" class="btn">Gallery</a><br>
        
        
        
        <a href="/cellxgene/posts/demo-data" class="btn">Demo datasets</a><br>
        
        
        
        <a href="/cellxgene/posts/prepare" class="btn">Preparing your data</a><br>
        
        
        
        <a href="/cellxgene/posts/launch" class="btn">Launching cellxgene</a><br>
        
        
        
        <a href="/cellxgene/posts/hosted" class="btn">Hosting cellxgene</a><br>
        
        
        
        <a href="/cellxgene/posts/annotations" class="btn">Annotating data</a><br>
        
        
        
        <a href="/cellxgene/posts/methods" class="btn">Methods</a><br>
        
        
        
        <a href="/cellxgene/posts/troubleshooting" class="btn"><b>Troubleshooting</b></a><br>
        
        
        
        <a href="/cellxgene/posts/roadmap" class="btn">Roadmap</a><br>
        
        
        
        <a href="/cellxgene/posts/contribute" class="btn">Contributing (ideas or code)</a><br>
        
        
        
        <a href="/cellxgene/posts/contact" class="btn">Contact & finding help</a><br>
        
        
        
        <a href="/cellxgene/posts/cellxgene_cziscience_com" class="btn">cellxgene.cziscience.com</a><br>
        
        

        <a href="https://github.com/chanzuckerberg/cellxgene" class="btn" target="_blank">Code</a>
    </p>
    </header>

      <section>
      <h1 id="troubleshooting-tips--tricks">Troubleshooting tips &amp; tricks</h1>

<h4 id="i-tried-to-pip-install-cellxgene-and-got-a-weird-error-i-dont-understand">I tried to <code class="language-plaintext highlighter-rouge">pip install cellxgene</code> and got a weird error I don’t understand</h4>

<p>This may happen, especially as we work out bugs in our installation process! Please create a new <a href="https://github.com/chanzuckerberg/cellxgene/issues">Github issue</a>, explain what you did, and include all the error messages you saw. It’d also be super helpful if you call <code class="language-plaintext highlighter-rouge">pip freeze</code> and include the full output alongside your issue.</p>

<h4 id="i-have-a-big-dataset-how-can-i-make-cellxgene-run-as-fast-as-possible">I have a BIG dataset, how can I make cellxgene run as fast as possible?</h4>

<p>If your dataset requires gigabytes of disk space, you may need to select an appropriate storage format in order to effectively utilize <code class="language-plaintext highlighter-rouge">cellxgene</code>. Tips and tricks:</p>

<ul>
  <li><code class="language-plaintext highlighter-rouge">cellxgene</code> is optimized for columnar data access. For large datasets, format the expression matrix (<code class="language-plaintext highlighter-rouge">.X</code>) as either a <a href="https://docs.scipy.org/doc/scipy/reference/generated/scipy.sparse.csc_matrix.html">SciPy CSC sparse matrix</a> or a dense Numpy array (whichever creates a smaller <code class="language-plaintext highlighter-rouge">h5ad</code> file). If you are using <code class="language-plaintext highlighter-rouge">cellxgene prepare</code>, include the <code class="language-plaintext highlighter-rouge">--sparse</code> flag to ensure <code class="language-plaintext highlighter-rouge">.X</code> is formatted as a CSC sparse matrix (by default, <code class="language-plaintext highlighter-rouge">.X</code> will be a dense matrix).</li>
  <li>By default, <code class="language-plaintext highlighter-rouge">cellxgene</code> loads the dataset into memory, and start time is directly proportional to <code class="language-plaintext highlighter-rouge">h5ad</code> file size and the speed of your file system. Expect that large (e.g., million cell) datasets will take minutes to load, even on relatively fast computers with a high performance local hard drive. Once loaded, exploring metadata should still be quick. If this start time is a problem, try the <code class="language-plaintext highlighter-rouge">--backed</code> flag, which will attempt to lazily load data as needed (caveat: subsequent data access may be slower).</li>
  <li>If your dataset size exceeds the size of memory (RAM) on the host computer, differential expression calculations will be extremely slow (or fail, if you run out of virtual memory). In this case, we recommend running with the <code class="language-plaintext highlighter-rouge">--disable-diffexp</code> flag. For datasets that are extremely large, you may also find the <code class="language-plaintext highlighter-rouge">--backed</code> flag improves your ability to explore them.</li>
</ul>

<h4 id="im-following-the-developer-instructions-and-get-an-error-about-missing-files-and-directories-when-trying-to-build-the-client">I’m following the developer instructions and get an error about “missing files and directories” when trying to build the client</h4>

<p>This is likely because you do not have node and npm installed, we recommend using <a href="https://github.com/creationix/nvm">nvm</a> if you’re new to using these tools.</p>

      </section>
      <footer>
        
        <p>This project is maintained by <a href="https://github.com/chanzuckerberg">chanzuckerberg</a></p>
        
      </footer>
    </div>
    <script src="/cellxgene/assets/js/scale.fix.js"></script>
    
  </body>
</html>
